src/ZoneServer/World/Storage/Storage.cs

```diff
@@ -0,0 +1,691 @@
+
+				// No space in existing stacks and no new position available
+				if (this.CheckStorageFull() && stackAvailableAmount <= 0)
+					return 0;
+
+				// Existing stacks can't stack and adding to new position is disabled
+				if (!ZoneServer.Instance.Conf.World.StorageMultiStack && stackAvailableAmount <= 0)
+					return 0;
+
+				availableAmount = stackAvailableAmount;
+
+				// We can also add the stack to empty spaces if multistack is enabled
+				if (ZoneServer.Instance.Conf.World.StorageMultiStack && !this.CheckStorageFull())
+					availableAmount = item.Data.MaxStack;
+			}
+			else
+			{
+				// Checks if non stackable item can be placed in storage
+				if (this.CheckStorageFull())
+					return 0;
+
+				availableAmount = item.Amount;
+			}
+
+			return availableAmount;
+		}
+
+		/// <summary>
+		/// Adds the available amount of an item to storage.
+		/// Updates client.
+		/// </summary>
+		/// <param name="interactingChar">Character to update</param>
+		/// <param name="itemToStore">Item to store</param>
+		/// <param name="invType">Type of storage inventory</param>
+		/// <returns></returns>
+		private StorageResult AddItemAvailableAmount(Character interactingChar, Item itemToStore, InventoryType invType)
+		{
+			// For stackable items, add item to existing stacks.
+			if (itemToStore.IsStackable)
+			{
+				this.TryGetStackableItems(itemToStore, out var existingStorageItems);
+				var addedAmount = 0;
+				foreach (var existingStorageItem in existingStorageItems)
+				{
+					var stackedAmount = this.AddItemStack(existingStorageItem.Value, itemToStore.Amount - addedAmount);
+					addedAmount += stackedAmount;
+
+					if (stackedAmount > 0)
+						Send.ZC_ITEM_ADD(interactingChar, existingStorageItem.Value, existingStorageItem.Key, stackedAmount, InventoryAddType.New, invType);
+				}
+
+				// If all items have been stored, return success
+				if (addedAmount >= itemToStore.Amount)
+					return StorageResult.Success;
+
+				// Residue
+				itemToStore.Amount -= addedAmount;
+			}
+
+			// Adds item to new position
+			var storageResult = this.Add(itemToStore, out var addedPosition);
+			if (storageResult != StorageResult.Success)
+				return storageResult;
+			Send.ZC_ITEM_ADD(interactingChar, itemToStore, addedPosition, itemToStore.Amount, InventoryAddType.New, invType);
+
+			return storageResult;
+		}
+
+		/// <summary>
+		/// Adds an item to storage and removes from given character.
+		/// Updates client.
+		/// </summary>
+		/// <param name="character">Character that is performing interaction</param>
+		/// <param name="objectId">ObjectId of item to store</param>
+		/// <param name="amount">Amount of item to store</param>
+		/// <param name="invType">Storage inventory type</param>
+		/// <returns></returns>
+		protected StorageResult StoreItem(Character character, long objectId, int amount, InventoryType invType)
+		{
+			var item = character.Inventory.GetItem(objectId);
```

Comment:
Would it not feel more natural to get the item you want to store and pass it in instead of getting it from the character in here? That way you could also add arbitrary items, and not only ones found on the character already.

Reply:
The other method in this class allows for that.
This method is just a simple shortcut to do so while also updating a character.